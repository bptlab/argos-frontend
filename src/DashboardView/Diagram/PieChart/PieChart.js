import React, { Component } from 'react';
import Chart from "chart.js"
import './PieChart.css';

const stateColor = {
    ERROR: "#D33F49",
    WARNING: "#EF9D6E",
    RUNNING: "#9FAB61",
};

class PieChart extends Component {
    getDisplayDataOf(chartData) {
        const displayData = [];
        const chartValueSum = chartData.reduce((pv, cv) => pv + cv, 0);
        chartData.forEach(function (item, index) {
            displayData.push(chartData[index] / chartValueSum);
        });
        return displayData;
    }

    getColors(chartLabels) {
        const backgroundColors = [];
        chartLabels.forEach(function (item) {
            backgroundColors.push(stateColor[item]);
        });
        return backgroundColors;
    }

    componentDidUpdate() {
        const charWrapper = this.refs.canvas;
        const chartContext = charWrapper.getContext('2d');
        const chart = new Chart(chartContext, {
            type: 'doughnut',
            options: {
                responsive: true
            },
            data: {
                labels: this.props.chartLabels,
                datasets: [{
                    data: this.getDisplayDataOf(this.props.chartData),
                    backgroundColor: this.getColors(this.props.chartLabels)
                }]
            }
        });
        this.state = {
            chart: chart
        };
    }

    componentWillUnmount() {
        this.state.chart.destroy();
    }

    render() {
        return (
            <canvas ref="canvas" id="pieChart"/>
        );
    }
}

export default PieChart;
