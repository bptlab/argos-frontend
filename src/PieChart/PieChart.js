import React, { Component } from 'react';
import './PieChart.scss';

class PieChart extends Component {
    render() {
        return (<div></div>
        );
    }
}

export default PieChart;
