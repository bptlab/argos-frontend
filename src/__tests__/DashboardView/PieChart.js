import React from 'react';
import PieChart from '../../DashboardView/Diagram/PieChart/PieChart';
import renderer from 'react-test-renderer';

const chartLabels = ["ERROR", "ERROR", "RUNNING"];
const chartData = [2, 3, 5];
let instance;

test("Correct drawing of PieChart", () => {
    const component = renderer.create(
        <PieChart
            ref={(child) => {instance = child}}
            chartData={chartLabels}
            chartLabels={chartData}
            onStateExcludeInput={() => {/* do nothing in case of test */}}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
}); 

test("Correct extraction of data for displaying", () => {
    const results = instance.getDisplayDataOf(chartData);
    expect(results).toEqual([0.2, 0.3, 0.5]);
});

test("Correct extraction of data for displaying", () => {
    const results = instance.getColors(chartLabels);
    expect(results).toEqual(["#D33F49", "#D33F49", "#9FAB61"]);
});