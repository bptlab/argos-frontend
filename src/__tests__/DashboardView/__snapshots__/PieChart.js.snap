exports[`test Correct drawing of PieChart 1`] = `
<canvas
  id="pieChart" />
`;
